src/aws-cpp-sdk-core/source/auth/GeneralHTTPCredentialsProvider.cpp

```diff
@@ -0,0 +1,204 @@
+          return false;
+        } else {
+          // full URI is not used in case of a relative one present
+          return true;
+        }
+    }
+
+    if (!absoluteUri.empty())
+    {
+        // If the resolved URI’s scheme is HTTPS, its hostname may be used in the request
+        if (Aws::Utils::StringUtils::ToLower(absoluteUri.c_str()).rfind(Aws::String("https://"), 0) == 0) // if starts_with
+        {
+            return true;
+        }
+
+        Aws::Http::URI absUri(absoluteUri);
+        const Aws::String& authority = absUri.GetAuthority();
+
+        // Otherwise, implementations MUST fail to resolve when the URI hostname does not satisfy any of the following conditions
+        // address is an ECS / EKS container host
+        if (authority == "169.254.170.2") // ECS container host
+        {
+            return true;
+        } else if (authority == "169.254.170.23" || absUri.GetAuthority() == "fd00:ec2::23") { // EKS container host
+            return true;
+        }
+
+        // address is within the loop back
+        // IPv4
+        if (authority.rfind(Aws::String("127.0.0."), 0) == 0 && authority.size() > 8 && authority.size() <= 11)
+        {
+            Aws::String octet = authority.substr(8);
+            if ((octet.find_first_not_of("0123456789") != Aws::String::npos) ||
+                (Aws::Utils::StringUtils::ConvertToInt32(octet.c_str()) > 255)) {
+                AWS_LOGSTREAM_WARN(GEN_HTTP_LOG_TAG, "Can't use General HTTP Provider: AWS_CONTAINER_CREDENTIALS_FULL_URI ip address is malformed: " << authority);
+                return false;
+            } else {
+                return true;
+            }
+        }
+
+        // IPv6
+        if (authority == "::1" || authority == "0:0:0:0:0:0:0:1" || authority == "[::1]" || authority == "[0:0:0:0:0:0:0:1]")
+        {
+            return true;
+        }
+
+        AWS_LOGSTREAM_WARN(GEN_HTTP_LOG_TAG, "Can't use General HTTP Provider: AWS_CONTAINER_CREDENTIALS_FULL_URI is not HTTPS and is not within loop back CIDR: " << authority);
+        return false;
+    }
+
+    // both relativeUri and absoluteUri are empty
+    return false;
+}
+
+GeneralHTTPCredentialsProvider::GeneralHTTPCredentialsProvider(const Aws::String& relativeUri,
+                                                               const Aws::String& absoluteUri,
+                                                               const Aws::String& authToken,
+                                                               const Aws::String& authTokenFilePath,
+                                                               long refreshRateMs) :
+    m_authTokenFilePath(authTokenFilePath),
+    m_loadFrequencyMs(refreshRateMs)
+{
+    if (ShouldCreateGeneralHTTPProvider(relativeUri, absoluteUri, authToken))
+    {
+        AWS_LOGSTREAM_INFO(GEN_HTTP_LOG_TAG, "Creating GeneralHTTPCredentialsProvider with refresh rate " << refreshRateMs);
+        if (!relativeUri.empty()) {
```

Comment:
should we move this to a factory function and move the ecs client into the initializer list and the logic whether or not we should create a client into the factory function?

Reply:
I thought about it and even implemented this way on the first attempt.
But 
1. it won't put these checks on existing use-cases.
2. or will have to perform these checks twice.